### ISSUE TITLE
datadir can't be outside of prefix

### ISSUE THREAD

**Issue by Keruspe** [mesonbuild/meson#1398]
Hi,

Is there any reason for disallowing it? It worked fine with 0.37.0 and fails at least with 0.38.1.
At exherbo, we put all arch-specific stuff into /usr/<insert the arch name here> and the non-arch-specific stuff into e.g. /usr/share, so the datadir is outside of prefix.

As a workaround, setting datadir to "../data" makes it work, but that feels really hacky

**Comment by Keruspe** [comment]
Maybe you want something like autotools which is separating on one side prefix and on the other side exec_prefix?

**Comment by nirbheek** [comment]
Could you describe your layout a bit more? What are the complete paths for `libdir`, `datadir`, `bindir`, etc? What value are you using for `prefix`?

The reason why this was done is: https://github.com/mesonbuild/meson/issues/1299#issuecomment-272922147

**Comment by Keruspe** [comment]
Here is an example of the standard arguments we pass to meson

```
meson                                                             \
        --prefix=/usr/$(exhost --target)                                  \
        --libdir=/usr/$(exhost --target)/lib                              \
        --libexecdir=/usr/$(exhost --target)/libexec                      \
        --includedir=/usr/$(exhost --target)/include                      \
        --sysconfdir=/etc                                                 \
        --datadir=/usr/share                                                \
        --mandir=/usr/share/man                                             \
        --default-library=shared                                          \
```

`$(exhost --target)` returns e.g `x86_64-pc-linux-gnu`

With the last release of meson we had to change the occurences of `/usr/share` with `../share` to make things work as expected.

**Comment by Keruspe** [comment]
I guess we could set prefix to /usr as we already specify explicitely the other dirs anyway, but it still feels weird compared to other build system, especially when relative path allows you to go to parents dirs, feels like a lack of consistency.

Or maybe `..` should be blacklisted then?

**Comment by nirbheek** [comment]
Yes, we should blacklist `..` as a path component. And yes, setting `prefix` to `/usr` makes sense in your case.

Also, this seems like a good excuse to add separate `exec_prefix` and `prefix`. We haven't added it so far because we couldn't find anyone who wanted it. It's going to interesting to implement it though, because right now everyone assumes you should `join_paths([prefix, bindir])`.

Ideally we want to implement #1081 and warn when people do `join_paths` with the wrong prefix-type, and recommend they use `xxxdir_abs` instead.
